File: Chapter10/higher_order_function/old/Archiver.java
package mastering.kotlin.performance.chapter10.higher_order_function.old;

import java.io.File;

public class Archiver {
    private CompressionStrategy strategy;

    public CompressionStrategy getStrategy() {
        return strategy;
    }

    public void setStrategy(CompressionStrategy strategy) {
        this.strategy = strategy;
    }

    public File archive(File file) {
        return strategy.compress(file);
    }
}

File: Chapter10/higher_order_function/Archiver.kt
package mastering.kotlin.performance.chapter10.higher_order_function

import java.io.File

fun File.archive(strategy: () -> File): File = strategy()

fun main(args: Array<String>) {
    File("input.txt").archive { TODO() }.also {
        // do something with an archived file
    }
}